src/snapred/backend/service/CalibrationService.py

```diff
@@ -120,6 +125,27 @@ def calculatePixelGroupingParameters(self, runs: List[RunConfig], groupingFile:
                 raise
         return data
 
+    @FromString
+    def hasState(self, runId: str, version: str):
+        stateID, _ = self.localDataService._generateStateId(runId)
+        calibrationStatePath = self.localDataService._constructCalibrationPath(stateID)
+
+        if os.path.exists(calibrationStatePath):
+            recordPath: str = calibrationStatePath + "{}/v{}/CalibrationRecord.json".format(runId, version)
+            if os.path.exists(recordPath):
+                return True
+            else:
+                return False
+        else:
```

Comment:
most of this should live in the localdataservice as you are doing filesystem level things, a service should be agnostic of the filesystem implementation

Reply:
(members prepended with an underbar `_` are considered private and shouldnt be called outside their class usually)

Reply:
you could in theory replace this code with a None check on the return of [this](https://github.com/neutrons/SNAPRed/blob/next/src/snapred/backend/data/LocalDataService.py#L464)
Though you would still need the preliminary check to see if the state folder even exists.

Reply:
Should be fixed.... hopefully.